File: WoofIOS/SwiftPackages/Presentation/Profile/Tests/PresentationProfileTests/ViewModel/ProfileMessageViewModelTest.swift
import DataSourceFakes
import Models
import Nimble
import PresentationProfile
import Quick
import Swinject
import SwinjectAutoregistration
import TestUtils
import UtilsTestExtensions

final class ProfileMessageViewModelTest: QuickSpec {
    override class func spec() {
        var container: Container!
        var viewModel: ProfileMessageViewModel!
        var messagesDataSource: FakeMessagesDataSource!
        let dog = DogFactory().produce()

        Given("I open a dog profile") {
            beforeEach {
                container = Container().injectForViewModelTests()
                TimeAdvancingFactory(container).withTestSchedulers()
                messagesDataSource = container~>
                viewModel = container.resolve(ProfileMessageViewModel.self, argument: dog)
            }

            justBeforeEach {
                viewModel.onViewAppear()
            }

            Then("The message bar is empty and sending is disabled") {
                expect(viewModel.state) == ProfileMessageViewModel.State(text: "", toast: nil)
                expect(viewModel.state.isSendEnabled) == false
            }

            When("I type a message") {
                justBeforeEach {
                    viewModel.onTextChange("Wanna play fetch?")
                }

                Then("Sending is enabled") {
                    expect(viewModel.state.text) == "Wanna play fetch?"
                    expect(viewModel.state.isSendEnabled) == true
                }

                And("I tap send") {
                    justBeforeEach {
                        viewModel.onSendTap()
                        TimeAdvancingFactory(container).tick()
                    }

                    Then("The message is sent, the bar is cleared, and a confirmation is shown") {
                        expect(messagesDataSource.sentMessages) == [SentMessage(dogId: 1, text: "Wanna play fetch?")]
                        expect(viewModel.state) == ProfileMessageViewModel.State(
                            text: "",
                            toast: .messageSent(name: "Bruno")
                        )
                    }
                }

                And("Sending fails") {
                    beforeEach {
                        MessagesDataSourceFactory(container).withSendMessageError()
                    }

                    justBeforeEach {
                        viewModel.onSendTap()
                        TimeAdvancingFactory(container).tick()
                    }

                    Then("The typed message is kept and an error is emitted") {
                        expect(viewModel.state.text) == "Wanna play fetch?"
                        expect(viewModel.error).toNot(beNil())
                    }
                }
            }

            When("I send a blank message") {
                justBeforeEach {
                    viewModel.onTextChange("   ")
                    viewModel.onSendTap()
                    TimeAdvancingFactory(container).tick()
                }

                Then("Nothing is sent and an empty message error is emitted") {
                    expect(messagesDataSource.sentMessages) == []
                    expect(viewModel.error) == .emptyMessage
                }
            }
        }
    }
}

File: WoofAndroid/presentation/profile/src/test/kotlin/com/perrystreet/woof/presentation/profile/viewmodel/ProfileMessageViewModelTest.kt
package com.perrystreet.woof.presentation.profile.viewmodel

import com.perrystreet.woof.datasource.messages.FakeMessagesDataSource
import com.perrystreet.woof.datasource.messages.MessagesDataSourceFactory
import com.perrystreet.woof.models.errors.MessageException
import com.perrystreet.woof.presentation.profile.factory.DogFactory
import com.perrystreet.woof.presentation.profile.spec.ViewModelBehaviorSpec
import com.perrystreet.woof.presentation.profile.uimodel.ProfileToastUIModel
import com.perrystreet.woof.presentation.profile.utils.UiObservableExtensions.test
import com.perrystreet.woof.testutils.rx.lastValue
import com.perrystreet.woof.testutils.scheduler.TimeAdvancingFactory
import org.amshove.kluent.shouldBeEqualTo
import org.koin.core.parameter.parametersOf
import org.koin.test.inject

class ProfileMessageViewModelTest : ViewModelBehaviorSpec() {
    private val dog = DogFactory().produce()
    private val viewModel: ProfileMessageViewModel by inject { parametersOf(dog) }
    private val messagesDataSource: FakeMessagesDataSource by inject()

    init {
        Given("I open a dog profile") {
            TimeAdvancingFactory().withTestSchedulers()

            beforeEach {
                viewModel.onViewAppear()
            }

            Then("The message bar is empty and sending is disabled") {
                val state = viewModel.state.test().lastValue()
                state shouldBeEqualTo ProfileMessageViewModel.State(text = "", toast = null)
                state.isSendEnabled shouldBeEqualTo false
            }

            When("I type a message") {
                beforeEach {
                    viewModel.onTextChange("Wanna play fetch?")
                }

                Then("Sending is enabled") {
                    val state = viewModel.state.test().lastValue()
                    state.text shouldBeEqualTo "Wanna play fetch?"
                    state.isSendEnabled shouldBeEqualTo true
                }

                And("I tap send") {
                    beforeEach {
                        viewModel.onSendTap()
                        TimeAdvancingFactory().tick()
                    }

                    Then("The message is sent, the bar is cleared, and a confirmation is shown") {
                        messagesDataSource.sentMessages shouldBeEqualTo listOf(1L to "Wanna play fetch?")
                        viewModel.state.test().lastValue() shouldBeEqualTo ProfileMessageViewModel.State(
                            text = "",
                            toast = ProfileToastUIModel.MessageSent("Bruno"),
                        )
                    }
                }

                And("Sending fails") {
                    MessagesDataSourceFactory().withSendMessageError()

                    beforeEach {
                        viewModel.onSendTap()
                        TimeAdvancingFactory().tick()
                    }

                    Then("The typed message is kept and an error is emitted") {
                        viewModel.state.test().lastValue().text shouldBeEqualTo "Wanna play fetch?"
                        viewModel.error.test().lastValue().isPresent shouldBeEqualTo true
                    }
                }
            }

            When("I send a blank message") {
                beforeEach {
                    viewModel.onTextChange("   ")
                    viewModel.onSendTap()
                    TimeAdvancingFactory().tick()
                }

                Then("Nothing is sent and an empty message error is emitted") {
                    messagesDataSource.sentMessages shouldBeEqualTo emptyList()
                    viewModel.error.test().lastValue().get() shouldBeEqualTo MessageException.EmptyMessage
                }
            }
        }
    }
}
